Support Ethernet on SAM E5x

stephanosio opened the pull request:
This PR includes a series of commits to support the GMAC Ethernet peripheral on the Atmel SAM0 family devices, and enable Ethernet support on the Atmel SAM E53 and E54 SoCs.

Closes #23001.

NOTE: This patch has been tested and verified working on the `atsame54_xpro` board using the `samples/net/dhcpv4_client` sample.

stephanosio committed:
dts: atmel: Add SAM E5x GMAC instance

This commit adds a GMAC instance to the SAM E5x device tree, along with
the refactoring necessary to specify the SAM E5x-specific components.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: atsame54_xpro: Enable GMAC in device tree

This commit enables the GMAC (Ethernet) instance in the `atsame54_xpro`
device tree.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
drivers: pinmux: Add more pin function definitions

This commit adds more pin function definitions (PINMUX_FUNC_I through
PINMUX_FUNC_P) to the pinmux interface header.

The SAM D5x and E5x series devices, for instances, define up to the
"function N" and this change is necessary to support such devices.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: atmel_sam0: Add GMAC fix-up for SAM D/E5x

This commit adds the MCLK clock configuration symbol fix-up for the
GMAC peripheral.

The APB-agnostic clock configuration fix-up symbols map to the
SoC-specific APB, in order to accommodate different SoC variants with
the GMAC on different APBs.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
drivers: ethernet: eth_sam_gmac: Add SAM0 family support

This commit adds the GMAC driver support for the Ethernet-capable SAM0
family devices (SAM E53 and E54 at this time).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: arm: atsame54_xpro: Support Ethernet

This commit enables the GMAC Ethernet support for the `atsame54_xpro`
board which includes an on-board RMII PHY.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

pfalcon reviewed (commented):
.

stephanosio commented:
Rebased

galak commented on boards/arm/atsame54_xpro/pinmux.c:
```diff
@@ -106,6 +106,20 @@ static int board_pinmux_init(struct device *dev)
 	pinmux_pin_set(muxa, 24, PINMUX_FUNC_H);
 #endif
+
+#if DT_ATMEL_SAM_GMAC_42000800_BASE_ADDRESS
```
Use a node label

stephanosio replied:
@galak Sure, I will do that.

Do you want me to convert the rest of such in this file as well? or are you already doing it somewhere else?

stephanosio replied:
I have converted this change to use the new DT macros, and did not touch the rest here assuming that will be done in a separate mass conversion PR.

galak commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -340,12 +358,17 @@ static inline void eth_sam_gmac_init_qav(Gmac *gmac)
 #endif
 
+#ifdef __DCACHE_PRESENT
 /*
```
Nit; can you move the cache helpers up so we only need one #ifdef __DCACHE_PRESENT / #else / #endif block?

stephanosio replied:
@nandojve Could you look into this since I took this commit from your PR?

nandojve replied:
Like this 9518832c.

stephanosio replied:
Rebased https://github.com/zephyrproject-rtos/zephyr/commit/9518832caafcd0d1f512ca542f7482a01a5a7bd9 into this PR.

Fixed.

galak commented on soc/arm/atmel_sam0/common/gmac_fixup_samd5x.h:
```diff
@@ -0,0 +1,22 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Can you add a comment about what these fixups are for/why they are needed.

stephanosio replied:
I will add the explanation in the commit message as in-line comment here.

stephanosio replied:
Done

galak reviewed (changes requested):
In general looks fine, some minor style comments.  Mostly want a comment about what gmac_fixup_samd5x.h is all about.

stephanosio commented:
@benpicco @nandojve Could you review this and, if there is no problem, approve this?

stephanosio committed:
drivers: ethernet: sam_gmac: Add SAM E54 maximum queue count reference

This commit adds a reference to the SAM E54 maximum queue count value
in the device tree for specifying the range of `ETH_SAM_GMAC_QUEUES`
configuration.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
Rebased, verified working on `atsame54_xpro`.

@galak Could you revisit this? Your reviews comments were addressed and your change request is still pending.

galak commented on drivers/ethernet/Kconfig.sam_gmac:
```diff
@@ -18,6 +18,7 @@ config ETH_SAM_GMAC_QUEUES
 	range 1 $(dt_node_int_prop_int,/soc/ethernet@40050088,num-queues) if SOC_SERIES_SAME70 || \
 									     SOC_SERIES_SAMV71
 	range 1 $(dt_node_int_prop_int,/soc/ethernet@40034000,num-queues) if SOC_SERIES_SAM4E
+	range 1 $(dt_node_int_prop_int,/soc/ethernet@42000800,num-queues) if SOC_SERIES_SAME54
```
I'd still prefer to remove this dt_node_int_prop_int
